typing/typecore.ml

```diff
@@ -711,41 +711,49 @@ end) = struct
+       Env.lookup_error loc' env' err
+
   (** [disambiguate] selects a concrete description for [lid] using
-     some contextual information: an optional [expected_type], and
-     a list of candidates [lbls]. If [expected_type] is [None], it
-     just returns the head of [lbls]. Otherwise, it extracts the
-     description directly from the corresponding type definition,
-     except for extension types, where it has to choose it inside
-     [lbls]. *)
-  (* Here [scope] contains the labels equal to [lid] in the current
-     environment, and [lbls] is a (potentially strict) subset of
-     [scope], see [disambiguate_label_by_ids] *)
-  let disambiguate ?(warn=Location.prerr_warning) ?scope
-                   usage lid env expected_type lbls =
-    let scope = match scope with None -> lbls | Some l -> l in
+     some contextual information:
+     - an optional [expected_type]
+     - a list of candidates labels in the current lexical scope,
+       [candidates_in_scope], that is actually at the type
+       [(label_descr list, lookup_error) result] so that the
+       lookup error is only raised when necessary
+     - a filtering criterion on candidates in scope [filter_candidates],
+       representing extra contextual information that can help
+       candidate selection (see [disambiguate_label_by_ids])
+   *)
+  let disambiguate ?(warn=Location.prerr_warning) usage lid env
+        expected_type
+        (filter_candidates : 'a list -> ('a list, 'a list) result)
+        candidates_in_scope =
     let lbl = match expected_type with
     | None ->
         (* no expected type => no disambiguation *)
-        begin match lbls with
-        | (Error(loc', env', err) : _ result) ->
-            Env.lookup_error loc' env' err
-        | Ok [] -> assert false
+        begin match filter_candidates (force candidates_in_scope) with
+        | Ok [] | Error []-> assert false
+        | Error((lbl, _use) :: _rest) -> lbl (* will fail later *)
         | Ok((lbl, use) :: rest) ->
             use ();
             warn_if_ambiguous warn lid env lbl rest;
             lbl
         end
     | Some(tpath0, tpath, principal) ->
-        (* first look for a disambiguation solution
-           in the current lexical scope *)
-        begin match disambiguate_by_type env tpath scope with
+        (* first look for a disambiguation solution in the current
+           lexical scope; note that [disambiguate_by_type] does not
+           force [candidates_in_scope]: we just skip this case if there
+           are no candidates in scope *)
+        begin match disambiguate_by_type env tpath candidates_in_scope with
         | lbl, use ->
           use ();
           if not principal then begin
             (* Check if non-principal type is affecting result *)
-            match lbls with
-            | (Error _ : _ result) | Ok [] -> warn_non_principal warn lid
+            match filter_candidates (force candidates_in_scope) with
```

Comment:
TODO: This case forces too much, the previous code would ignore the `Error` in this situation.

Reply:
(fixed)